src/main/java/com/linecorp/armeria/client/metrics/MetricCollectingClient.java

```diff
@@ -30,16 +33,25 @@
      * <p>It is generally recommended to define your own name for the service instead of using something like
      * the Java class to make sure otherwise safe changes like renames don't break metrics.
      */
-    public static Function<Client, Client> newDropwizardDecorator(MetricRegistry metricRegistry,
-                                                                  String metricNamePrefix) {
+    public static <I, O> Function<Client<I, O>, Client<I, O>> newDropwizardDecorator(
+            MetricRegistry metricRegistry, String metricNamePrefix) {
 
-        return client -> new MetricCollectingClient(
+        return client -> new MetricCollectingClient<>(
                 client,
                 new DropwizardMetricConsumer(metricRegistry, metricNamePrefix));
     }
 
+    private final MetricConsumer consumer;
+
+    public MetricCollectingClient(Client<?, ?> delegate, MetricConsumer consumer) {
+        super(delegate);
+        this.consumer = requireNonNull(consumer, "consumer");
+    }
 
-    public MetricCollectingClient(Client client, MetricConsumer metricConsumer) {
-        super(client, codec -> new MetricCollectingClientCodec(codec, metricConsumer), Function.identity());
+    @Override
+    public O execute(ClientRequestContext ctx, I req) throws Exception {
+        ctx.awaitRequestLog().thenAccept(consumer::onRequest);
+        ctx.awaitResponseLog().thenAccept(consumer::onResponse);
+        return delegate().execute(ctx, req);
```

Comment:
I like the idea. Let me:

- add `LogCollectingClient/Service` and
- make `LoggingClient/Service` and `MetricCollectingClient/Service` extend the `LogCollectingClient/Service`.

WDYT?